### DISCUSSION TITLE
Протокол I2C

### DISCUSSION THREAD

**Discussion by neitri** [dontsovcmc/waterius#191] [Ideas]
Рассматривал исходники и возник вопрос по поводу протокола обмена данными. 
Если разобраться с чтением i2c eeprom памяти то это выглядит следующим образом
START|DEV_ADDRESS|W|ADDRESS| STOP| START|DEV_ADDRESS|R|byte 0|.......|byte N|STOP
Т.е. всю микросхему можно прочитать 2мя запросами. 
1й запрос на запись передается адрес с какой ячейки читать(писать)
2й запрос на чтение. последовательно считывается необходимое количество байт.
Что мы имеем в ватериусе.
Каждый байт считывается отдельным запросом.
Т.е. для чтения 22байт передается еще 22е стартовые посылки длинной по 8 бит. 
START|DEV_ADDRESS|R|byte 0|STOP|START|DEV_ADDRESS|R|byte 1|STOP|START|DEV_ADDRESS|R|byte 2|STOP|...|START|DEV_ADDRESS|R|byte N|STOP
Может я не в курсе каких то особенностей тиньки или библиотеки Wire, но можно сократить время чтения данных примерно в 2 раза.
2й вопрос.
Отдельное чтение режима работы?
Можно включить в структуру SlaveData добавить режим работы и считывать его за один раз.

**Comment by dontsovcmc** [comment]
1. Зачем оптимизировать и так быструю i2c?
Команда 1 раз подается на чтение структуры и дальше она вычитывается. Там нечего оптимизировать.
2. А есть флаг mode - его и дополним.

**Reply by neitri** [reply]
Разобрался я как работает Wire.
Функция OnRequest вызывается при успешном сравнении адреса. Функция передает 1 байт.
В самой библиотеке буфер используется 16 байт. После его увеличения до 32 байт. Получилось передать весь буфер одной посылкой.
По времени получается 12мс если Start адрес + команда на чтение start адрес + 23 байта занимает около 12мс
Оригинальный запрашивая по 1 байту занимает 22мс.
Также и с чтением. Команды read и write не совсем читают и пишут данные. Они читают из буфера или пишут в буфер. requestfrom(адрес, количество байт) вычитывает заданное количество байт и только после этого командами read можно вычитать буфер. 
Зная как работает i2c это не совсем логично. 
Как итог. В прошивке тини можно отказаться от библиотеки Wire, а оставить библиотеку USI_TWI_Slave который обернут в Wire.
